main.go

```diff
@@ -136,9 +136,67 @@ func main() {
 		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
 		prometheus.NewGoCollector(),
 	)
-	go telemetryServer(ksmMetricsRegistry, opts.TelemetryHost, opts.TelemetryPort)
 
-	serveMetrics(ctx, kubeClient, storeBuilder, opts, opts.Host, opts.Port, opts.EnableGZIPEncoding)
+	var g run.Group
+
+	m := metricshandler.New(
+		opts,
+		kubeClient,
+		storeBuilder,
+		opts.EnableGZIPEncoding,
+	)
+	// Run MetricsHandler
+	{
+		ctxMetricsHandler, cancel := context.WithCancel(ctx)
+		g.Add(func() error {
+			return m.Run(ctxMetricsHandler)
+		}, func(error) {
+			cancel()
+		})
+	}
+
+	telemetryMux := buildTelemetryServer(ksmMetricsRegistry)
+	telemetryServer := http.Server{Handler: telemetryMux}
+	telemetryListenAddress := net.JoinHostPort(opts.TelemetryHost, strconv.Itoa(opts.TelemetryPort))
+	telemetryLn, err := net.Listen("tcp", telemetryListenAddress)
+	if err != nil {
+		klog.Fatalf("Failed to create Telemetry Listener: %v", err)
+	}
+	metricsMux := buildMetricsServer(kubeClient, storeBuilder, m, opts)
+	metricsServer := http.Server{Handler: metricsMux}
+	metricsServerListenAddress := net.JoinHostPort(opts.Host, strconv.Itoa(opts.Port))
+	metricsServerLn, err := net.Listen("tcp", metricsServerListenAddress)
+	if err != nil {
+		klog.Fatalf("Failed to create MetricsServer Listener: %v", err)
+	}
+
+	// Run Telemetry server
+	{
+		g.Add(func() error {
+			klog.Infof("Starting kube-state-metrics self metrics server: %s", telemetryListenAddress)
+			return telemetryServer.Serve(telemetryLn)
+		}, func(error) {
+			ctxShutDown, cancel := context.WithTimeout(ctx, 3*time.Second)
```

Comment:
What are we trying to achieve with this?

Reply:
The purpose of my WithTimeout was to let some time to gracefully shutdown but force a timeout if there was a long running query.

Reply:
that shouldn't be necessary, canceling allows the go routine to finish anyways

Reply:
Hmm, I understand.
Shutdown being a blocking call that can wait indefinitely, when would Cancel be called when using WithCancel?

Reply:
cancel for all run groups is called when any run group exists

Reply:
Oh ... then there is no reason to graceful shutdown and we can just Close it?

Reply:
Sorry my bad I understand now what this is achieving, I think it's worth trying this yes.

Reply:
Still, I understand your concern but I'm unsure on how to test whether this really does a Graceful shutdown or not.